File: iceoryx2/src/tracker/service.rs
// Copyright (c) 2025 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

use crate::config;
use crate::service::service_id::ServiceId;
use crate::service::static_config::StaticConfig;
use crate::service::{Service, ServiceDetails};
use iceoryx2_bb_elementary::CallbackProgression;
use std::collections::{HashMap, HashSet};

/// A tracker for monitoring services of a specific type.
///
/// The `Tracker` keeps track of services in the system, allowing for discovery
/// of new services and detection of services that are no longer available.
///
/// # Type Parameters
///
/// * `S` - The type of service to track, which must implement the `Service` trait
pub struct Tracker<S: Service> {
    services: HashMap<ServiceId, ServiceDetails<S>>,
}

impl<S: Service> Tracker<S> {
    /// Create a new Monitor instance.
    pub fn new() -> Self {
        Self {
            services: HashMap::new(),
        }
    }

    /// Synchronizes the tracker with the current state of services in the system.
    ///
    /// This method discovers all services of type `S` in the provided configuration,
    /// tracks new services, and removes services that are no longer available.
    ///
    /// # Arguments
    ///
    /// * `config` - The configuration to use for service discovery
    ///
    /// # Returns
    ///
    /// A tuple containing:
    /// * A set of service IDs of the newly discovered services, whose details are stored
    /// * A vector of static configurations for services, whose details are no longer stored
    pub fn sync(&mut self, config: &config::Config) -> (HashSet<ServiceId>, Vec<StaticConfig>) {
        let mut discovered_ids = HashSet::<ServiceId>::new();
        let mut new_ids = HashSet::<ServiceId>::new();

        let _ = S::list(config, |service| {
            let id = service.static_details.service_id();
            discovered_ids.insert(id.clone());

            // Track new services.
            if !self.services.contains_key(id) {
                new_ids.insert(id.clone());
                self.services.insert(id.clone(), service);
            }
            CallbackProgression::Continue
        });

        // Get the details of the services not discovered
        let undiscovered_ids: HashSet<ServiceId> = self
            .services
            .keys()
            .filter(|&id| !discovered_ids.contains(id))
            .cloned()
            .collect();
        let undiscovered_details = self.get_many(&undiscovered_ids);

        // Remove the services that were not discovered
        self.services.retain(|id, _| discovered_ids.contains(id));

        (new_ids, undiscovered_details)
    }

    /// Returns the static details of a single service with the given ID.
    ///
    /// # Arguments
    ///
    /// * `id` - The service ID to retrieve details for
    ///
    /// # Returns
    ///
    /// An option containing the static details if the service was found, or None otherwise
    pub fn get_one(&self, id: &ServiceId) -> Option<StaticConfig> {
        self.services
            .get(id)
            .map(|service| service.static_details.clone())
    }

    /// Returns the static details of the services with the given IDs.
    ///
    /// # Arguments
    ///
    /// * `ids` - A set of service IDs to retrieve details for
    ///
    /// # Returns
    ///
    /// A vector of static configurations for the services that were found
    pub fn get_many<'a, I>(&self, ids: I) -> Vec<StaticConfig>
    where
        I: IntoIterator<Item = &'a ServiceId>,
    {
        ids.into_iter()
            .filter_map(|id| {
                self.services
                    .get(id)
                    .map(|service| service.static_details.clone())
            })
            .collect()
    }
}

File: iceoryx2-cxx/src/service.cpp
// Copyright (c) 2024 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

#include "iox2/service.hpp"
#include "iox2/iceoryx2.h"
#include "iox2/internal/iceoryx2.hpp"
#include "iox2/service_details.hpp"
#include "iox2/service_type.hpp"
#include "iox2/static_config.hpp"

namespace iox2 {
template <ServiceType S>
auto Service<S>::does_exist(const ServiceName& service_name,
                            const ConfigView config,
                            const MessagingPattern messaging_pattern)
    -> iox2::legacy::expected<bool, ServiceDetailsError> {
    bool does_exist_result = false;
    auto result = iox2_service_does_exist(iox2::bb::into<iox2_service_type_e>(S),
                                          service_name.as_view().m_ptr,
                                          config.m_ptr,
                                          iox2::bb::into<iox2_messaging_pattern_e>(messaging_pattern),
                                          &does_exist_result);

    if (result == IOX2_OK) {
        return iox2::legacy::ok(does_exist_result);
    }

    return iox2::legacy::err(iox2::bb::into<ServiceDetailsError>(result));
}

template <ServiceType S>
auto Service<S>::details(const ServiceName& service_name,
                         const ConfigView config,
                         const MessagingPattern messaging_pattern)
    -> iox2::legacy::expected<iox2::legacy::optional<ServiceDetails<S>>, ServiceDetailsError> {
    iox2_static_config_t raw_static_config;
    bool does_exist = false;

    auto result = iox2_service_details(iox2::bb::into<iox2_service_type_e>(S),
                                       service_name.as_view().m_ptr,
                                       config.m_ptr,
                                       iox2::bb::into<iox2_messaging_pattern_e>(messaging_pattern),
                                       &raw_static_config,
                                       &does_exist);

    if (result != IOX2_OK) {
        return iox2::legacy::err(iox2::bb::into<ServiceDetailsError>(result));
    }

    if (!does_exist) {
        return iox2::legacy::ok(iox2::legacy::optional<ServiceDetails<S>>());
    }

    return iox2::legacy::ok(
        iox2::legacy::optional<ServiceDetails<S>>(ServiceDetails<S> { StaticConfig(raw_static_config) }));
}

template <ServiceType S>
auto list_callback(const iox2_static_config_t* const static_config, void* ctx) -> iox2_callback_progression_e {
    auto callback = static_cast<iox2::legacy::function<CallbackProgression(ServiceDetails<S>)>*>(ctx);
    auto result = (*callback)(ServiceDetails<S> { StaticConfig(*static_config) });
    return iox2::bb::into<iox2_callback_progression_e>(result);
}

template <ServiceType S>
auto Service<S>::list(const ConfigView config,
                      const iox2::legacy::function<CallbackProgression(ServiceDetails<S>)>& callback)
    -> iox2::legacy::expected<void, ServiceListError> {
    auto mutable_callback = callback;
    auto result = iox2_service_list(
        iox2::bb::into<iox2_service_type_e>(S), config.m_ptr, list_callback<S>, &mutable_callback);

    if (result == IOX2_OK) {
        return iox2::legacy::ok();
    }

    return iox2::legacy::err(iox2::bb::into<ServiceListError>(result));
}

template class Service<ServiceType::Ipc>;
template class Service<ServiceType::Local>;
} // namespace iox2
